45.md

```diff
@@ -0,0 +1,212 @@
+- `marketplace` - clientside software for searching `stalls` and purchasing `products`
+
+## Nostr Marketplace Clients
+
+### Merchant admin
+
+Where the `merchant` creates, updates and deletes `stalls` and `products`, as well as where they manage sales, payments and communication with `customers`.
+
+The `merchant` admin software can be purely clientside, but for `convenience` and uptime, implementations will likely have a server client listening for NOSTR events.
+
+### Marketplace
+
+`Marketplace` software should be entirely clientside, either as a stand-alone app, or as a purely frontend webpage. A `customer` subscribes to different merchant NOSTR public keys, and those `merchants` `stalls` and `products` become listed and searchable. The marketplace client is like any other ecommerce site, with basket and checkout. `Marketplaces` may also wish to include a `customer` support area for direct message communication with `merchants`.
+
+## `Merchant` publishing/updating products (event)
+
+A merchant can publish these events:
+| Kind    |                  | Description                                                                                                   | NIP                                     |
+|---------|------------------|---------------------------------------------------------------------------------------------------------------|-----------------------------------------|
+| `0    ` | `set_meta`       | The merchant description (similar with any `nostr` public key).                                               | [NIP01       ](https://github.com/nostr-protocol/nips/blob/master/01.md)                            |
+| `30017` | `set_stall`      | Create or update a stall.                                                                                     | [NIP33](https://github.com/nostr-protocol/nips/blob/master/33.md) (Parameterized Replaceable Event) |
+| `30018` | `set_product`    | Create or update a product.                                                                                   | [NIP33](https://github.com/nostr-protocol/nips/blob/master/33.md) (Parameterized Replaceable Event) |
+| `4    ` | `direct_message` | Communicate with the customer. The messages can be plain-text or JSON. | [NIP09](https://github.com/nostr-protocol/nips/blob/master/09.md)                                            |
+| `5    ` | `delete`         | Delete a product or a stall.                                                                                  | [NIP05](https://github.com/nostr-protocol/nips/blob/master/05.md)                                   |
+
+### Event `30017`: Create or update a stall.
+
+**Event Content**:
+```json
+{
+    "id": <String, UUID generated by the merchant. Sequential IDs (`0`, `1`, `2`...) are discouraged>,
+    "name": <String, stall name>,
+    "description": <String (optional), stall description>,
+    "currency": <String, currency used>,
+    "shipping": [
```

Comment:
This feels very specific and inflexible. What if the items require a more complex shipping cost structure, such as a weight/country matrix, cart subtotal quota, etc? Realistically, the current shipping option does not work for most sellers.

I would suggest `shipping` be optional, a simple string value describing the shipping costs. 

Then, I would add a `checkoutData` array similar in function to https://github.com/lnurl/luds/blob/luds/18.md
For example:
```json
    "checkoutData": {
        "name": { mandatory: boolean, description: string },
        "address": { mandatory: boolean, description: string },
        "country": { mandatory: boolean, description: string }
    }
```

And then, in the `customer` order event, add the following to the payload:
```json
   "checkoutData": { "name" : "Kukks", "address": "17, Chocolate Street", "country": "Russia"}
```


This would also omit the `shipping_id` and `address` from customer order event

Reply:
> What if the items require a more complex shipping cost structure, such as a weight/country matrix
 - I agree it does not cover complex shipping scenarios (was not intended to)
 - the solution (as far as this NIP goes) is for the merchant to send over DM to the customer any relevant information (like extra shipping cost for higher weights)
 - use-cases for : gift cards, promotions, discounts (over a certain quantity), etc are not included in this NIP (but can be part of future NIPs)

> I would suggest shipping be optional, a simple string value describing the shipping costs.
 - the shipping cost can vary depending on the shipping zone (so more costs are possible)
 - the `shipping zone id` is required, as the merchant cannot rely on the value (cost) sent by the client